src/emscatter.jl

```diff
@@ -0,0 +1,307 @@
+# 
+# For simplicity, we start with only the $m$-th component and take it as the incident part: 
+#
+# ```math
+# H_{inc}=J_m(kr).
+# ```
+#
+# For the scattered field, since the scattered wave should be going out, we can then expand it in terms of the Hankel function of the first kind (outgoing and incoming cylindrical waves are Hankel functions of the first and second kind [6]): 
+#
+# ```math
+# H_1=\alpha_mH_m^1(kr).
+# ```
+#
+# For the fields inside the cylinder, we require the field to be finite at $r=0$, which then constrains the field to be only the expansion of the Bessel functions: 
+#
+# ```math
+# H_2=\beta_mJ_m(nkr),
+# ```
+# 
+# where $n=\sqrt{\varepsilon}$ is the refractive index. 
+# 
+# Applying the boundary conditions (tangential part of the electri and magnetic field to be continuous): 
+#
+# ```math
+# H_{inc}+H_1=H_2|_{r=R},
+# ```
+#
+# ```math
+# \frac{\partial H_{inc}}{\partial r}+\frac{\partial H_1}{\partial r}=\frac{1}{\epsilon}\frac{\partial H_2}{\partial r}|_{r=R}.
+# ```
+# 
+# After some math, we get: 
+#
+# ```math
+# \alpha_m=\frac{J_m(nkR)J_m(kR)^\prime-\frac{1}{n}J_m(kR)J_m(nkR)^\prime}{\frac{1}{n}H_m^1(kR)J_m(nkr)^\prime-J_m(nkr)H_m^1(kr)^\prime},
+# ```
+# 
+# ```math
+# \beta_m = \frac{H_m^1(kR)J_m(kR)^\prime-J_m(kR)H_m^1(kR)^\prime}{\frac{1}{n}J_m(nkR)^\prime H_m^1(kR)-J_m(nkR)H_m^1(kR)^\prime},
+# ```
+# 
+# where $^\prime$ denotes the derivative, and the derivatives of the Bessel functions are obtained with the recurrent relations: 
+#
+# ```math
+# Y_m(z)^\prime=\frac{Y_{m-1}(z)-Y_{m+1}(z)}{2}
+# ```
+#
+# where $Y_m$ denotes any Bessel functions (Hankel functions).
+#
+# 
+# Finally, the analytical field is ($1/2k$ is the amplitude that comes from the unit line source excitation):
+# ```math
+# H(r>R)=\frac{1}{2k}\sum_m\left[\alpha_mi^mH_m^1(kr)+J_m(kr)\right]e^{im\theta}
+# ```
+#
+# ```math
+# H(r<=R)=\frac{1}{2k}\sum_m\beta_mi^mJ_m(nkr)e^{im\theta}
+# ```
+# 
+
+using SpecialFunctions
+dbesselj(m,z) = (besselj(m-1,z)-besselj(m+1,z))/2
+dhankelh1(m,z)= (hankelh1(m-1,z)-hankelh1(m+1,z))/2
+α(m,n,z) = (besselj(m,n*z)*dbesselj(m,z)-1/n*besselj(m,z)*dbesselj(m,n*z))/(1/n*hankelh1(m,z)*dbesselj(m,n*z)-besselj(m,n*z)*dhankelh1(m,z))
+β(m,n,z) = (hankelh1(m,z)*dbesselj(m,z)-besselj(m,z)*dhankelh1(m,z))/(1/n*dbesselj(m,n*z)*hankelh1(m,z)-besselj(m,n*z)*dhankelh1(m,z))
+
+function H_t(x,xc,r,ϵ,λ)
+    n = √ϵ
+    k = 2*π/λ
+    θ = angle(x[1]-xc[1]+1im*(x[2]-xc[2]))+π
+    M = 20
+    H0 = 0
+    if norm([x[1]-xc[1],x[2]-xc[2]])<=r
+        for m=-M:M
+            H0 += β(m,n,k*r)*cis(m*θ)*besselj(m,n*k*norm([x[1]-xc[1],x[2]-xc[2]]))
+        end
+    else
+        for m=-M:M
+            H0 += α(m,n,k*r)*cis(m*θ)*hankelh1(m,k*norm([x[1]-xc[1],x[2]-xc[2]]))+cis(m*θ)*besselj(m,k*norm([x[1]-xc[1],x[2]-xc[2]]))
+        end
+    end
+    return 1im/(2*k)*H0
+end
+# Construct the analytical solution with finite element functions
+a_theory(u,v) = ∫(u*v)*dΩ
```

Comment:
Why do you need to do this L2 projection?  A sentence clarifying this in the text would be very welcome.